Add stubs for URLPattern API

jelee53 opened the pull request:
#### b67afb5ae3430b50e1ce14800dc30cea8a281116
<pre>
Add stubs for URLPattern API
<a href="https://bugs.webkit.org/show_bug.cgi?id=280795">https://bugs.webkit.org/show_bug.cgi?id=280795</a>
<a href="https://rdar.apple.com/137166466">rdar://137166466</a>

Reviewed by Chris Dumez.

URLPattern API spec <a href="https://urlpattern.spec.whatwg.org.">https://urlpattern.spec.whatwg.org.</a>

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp: Added.
(WebCore::URLPattern::create):
(WebCore::URLPattern::test const):
(WebCore::URLPattern::exec const):
* Source/WebCore/Modules/url-pattern/URLPattern.h: Added.
* Source/WebCore/Modules/url-pattern/URLPattern.idl: Added.
* Source/WebCore/Modules/url-pattern/URLPatternInit.h: Added.
* Source/WebCore/Modules/url-pattern/URLPatternInit.idl: Added.
* Source/WebCore/Modules/url-pattern/URLPatternOptions.h: Added.
* Source/WebCore/Modules/url-pattern/URLPatternOptions.idl: Added.
* Source/WebCore/Modules/url-pattern/URLPatternResult.h: Added.
* Source/WebCore/Modules/url-pattern/URLPatternResult.idl: Added.
* Source/WebCore/Scripts/GenerateSettings.rb:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: <a href="https://commits.webkit.org/284734@main">https://commits.webkit.org/284734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a1ab691c15ddc220e08ba5ab72301b30a0c0bd7e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70292 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49698 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23056 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74381 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21470 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72409 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57498 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21310 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55716 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14193 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73358 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45227 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60606 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36178 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41886 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18041 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19831 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/63413 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63822 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18385 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76100 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69539 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14518 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17621 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63434 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14560 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60672 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63371 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15585 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11408 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5033 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91320 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45500 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/267 "") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/19904 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46574 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47851 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46316 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+ {
+ 	(void) input;  // Need to add this so build won't complain about unused variables
+ 	(void) baseURL;
+    (void) options;
```
You could silence the warning by not specifying the names, e.g. 
```
Ref<URLPattern> URLPattern::create(const URLPatternInput&, const String&, URLPatternOptions&&)
{
    return adoptRef(*new URLPattern());
}
```

cdumez commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -820,6 +820,11 @@ set(WebCore_NON_SVG_IDL_FILES
     Modules/webdatabase/SQLTransactionCallback.idl
     Modules/webdatabase/SQLTransactionErrorCallback.idl
 
+    Modules/url-pattern/URLPattern.idl
```
Needs to be alphabetically sorted.

cdumez commented on Source/WebCore/DerivedSources.make:
```diff
@@ -646,6 +646,10 @@ JS_BINDING_IDLS := \
     $(WebCore)/Modules/streams/WritableStreamDefaultWriter.idl \
     $(WebCore)/Modules/streams/WritableStreamSink.idl \
     $(WebCore)/Modules/storage/StorageManager.idl \
+	$(WebCore)/Modules/url-pattern/URLPattern.idl \
```
indentation issue.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
Wrong year

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
```
This should look like:
```
#include "config.h"
#include "URLPattern.h"

#include "URLPatternInit.h"
#include "URLPatternOptions.h"
#include "URLPatternResult.h"
#include <wtf/RefCounted.h>
```

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+
+ Ref<URLPattern> URLPattern::create(const URLPatternInput input, const String&& baseURL, URLPatternOptions&& options)
+ {
+ 	(void) input;  // Need to add this so build won't complain about unused variables
```
`UNUSED_PARAM(input);`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+{
+    (void) options;
+
+     if (input.has_value())
```
Please omit for now.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+
+ Ref<URLPattern> URLPattern::create(std::optional<URLPatternInput> input, URLPatternOptions&& options)
+{
+    (void) options;
```
UNUSED_PARAM()

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+ExceptionOr<bool> URLPattern::test(std::optional<URLPatternInput> input, std::optional<String>&& baseURL)
+{
+    
+	if (input.has_value())
```
Please omit for now.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+	}
+
+
+	if (baseURL.has_value())
```
Please omit for now.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+	}
+
+    return Exception { ExceptionCode::NotSupportedError };
+    // return false;
```
Please don't leave commented out code like this.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+
+void URLPattern::exec(std::optional<String>&& input, std::optional<String>&& baseURL)
+{
+    if (input.has_value())
```
Please omit for now.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+        ;
+    }
+
+    if (baseURL.has_value())
```
Please omit for now.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,96 @@
+        ;
+    }
+
+    // return std::nullopt;
```
No commented out code.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,95 @@
+// JBG TODO notes
```
Please remove all these comments.

jelee53 replied:
Absolutely! This is just a reference branch / draft for myself + anyone who wants to see my WiP! Definitely don't intend to put these comments in main!

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,95 @@
+
+
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
Wrong year.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,95 @@
+    ExceptionOr<bool> hasRegExpGroups() const { return Exception { ExceptionCode::NotSupportedError }; }
+
+private:
+
```
Please remove blank line here.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,95 @@
+
+    void exec(std::optional<String>&& input, std::optional<String>&& baseURL);
+
+    ExceptionOr<String> protocol() const { return Exception { ExceptionCode::NotSupportedError }; }
```
Seems like `return m_protocol;` would be shorter and more correct.

same below.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternInit.h:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
Wrong year.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternInit.idl:
```diff
@@ -0,0 +1,41 @@
+    JSGenerateToJSObject,
+    JSGenerateToNativeObject
+] dictionary URLPatternInit {
+  USVString protocol;
```
Not enough indentation.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternOptions.idl:
```diff
@@ -0,0 +1,30 @@
+// https://urlpattern.spec.whatwg.org
+
+dictionary URLPatternOptions {
+  boolean ignoreCase = false;
```
Please use 4-space indentation.

cdumez commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 70;
```
Please revert this version change.

jelee53 replied:
Oops this didn't get fixed in this second commit. Let me re-commit with this change.

jelee53 replied:
Actually the project won't build right now because I accidentally didn't add the pbxproj to the commit. Apologies. Let me re-generate & re-add.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,81 @@
+
+    std::optional<URLPatternResult> exec(std::optional<URLPatternInput>&&, std::optional<String>&& baseURL);
+
+    String protocol() const { return m_protocol; }
```
Should return a `const String&` to avoid copy when possible.

Ditto for all getters below.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,81 @@
+    static Ref<URLPattern> create(std::optional<URLPatternInput>, URLPatternOptions&&);
+    ~URLPattern();
+
+    ExceptionOr<bool> test(std::optional<URLPatternInput>, std::optional<String>&& baseURL);
```
Can the function be marked as `const`?

jelee53 replied:
Actually I don't think this can be marked as const because the class will be modified. 

One of the parameters is a String which presumably the class will parse and initialize member properties with.

cdumez replied:
> the class will be modified.

Will it? Sounds unlikely to me.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,81 @@
+
+    ExceptionOr<bool> test(std::optional<URLPatternInput>, std::optional<String>&& baseURL);
+
+    std::optional<URLPatternResult> exec(std::optional<URLPatternInput>&&, std::optional<String>&& baseURL);
```
ditto.

jelee53 replied:
I think I will leave this one as non-const actually. I do think this will modify class properties, too.

cdumez replied:
Seems odd that test() / exec() functions would modify the class.

jelee53 replied:
Yes you are right! Sorry, I misread the spec. I will make this change. 

For these two functions,  URLPattern instance is being compared against another input. So that particular instance's properties won't change.

jelee53 replied:
(
I don't know why I thought this but my original thinking was that when the input string is passed in, URLPattern will try to parse/compare the input string with the components, eg. protocol, password, etc and that would somehow affect the instance's member properties. 

But the member properties will most likely stay the same and the input string components are probably in some temporary variable. 
)

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,81 @@
+    String m_hash;
+
+    bool m_hasRegExpGroups;
+    bool m_ignoreCase = false;
```
We prefer the `bool m_ignoreCase { false }` style.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,81 @@
+    String m_search;
+    String m_hash;
+
+    bool m_hasRegExpGroups;
```
Uninitialized.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.idl:
```diff
@@ -0,0 +1,56 @@
+    readonly attribute boolean hasRegExpGroups;
+};
+
+typedef (USVString or URLPatternInit or URLPattern) URLPatternCompatible;
```
Doesn't appear to be used?

jelee53 replied:
It was part of the IDL spec so I implemented it. In URLPattern case, I think its if javascript tries to directly update a dictionary? I'm not sure either.

https://urlpattern.spec.whatwg.org/#typedefdef-urlpatterncompatible

cdumez replied:
They might have, I don't see it being used in the spec either. I would omit for now.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternOptions.h:
```diff
@@ -0,0 +1,34 @@
+namespace WebCore {
+
+struct URLPatternOptions {
+    bool ignoreCase = false;
```
`{ false }`

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,68 @@
+    UNUSED_PARAM(baseURL);
+
+    return adoptRef(*new URLPattern());
+}
```
You could add the param name when you start using it:
```
Ref<URLPattern> URLPattern::create(const URLPatternInput, const String&&, URLPatternOptions&&)
 {
     return adoptRef(*new URLPattern());
 }
```

cdumez replied:
Parameter names are useful, I would keep them.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternResult.idl:
```diff
@@ -0,0 +1,54 @@
+] dictionary URLPatternComponentResult {
+    USVString input;
+    // FIXME: Undefined in union is not supported.
+    // record<USVString, (undefined or USVString)> groups;
```
I just fixed this in https://commits.webkit.org/284691@main; could you try rebasing and uncomment this?

jelee53 replied:
It worked! Thanks Sihui!

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternResult.h:
```diff
@@ -0,0 +1,52 @@
+struct URLPatternComponentResult {
+    String input;
+    // FIXME: Undefined in union is not supported.
+    // using GroupsRecord = Vector<KeyValuePair<String, std::optional<String>>>;
```
This should be `GroupsRecord = Vector<KeyValuePair<String, std::variant<std::nullptr_t, String>>>;`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,80 @@
+public:
+    using URLPatternInput = std::variant<String, URLPatternInit>;
+
+    static Ref<URLPattern> create(const URLPatternInput, const String&& baseURL, URLPatternOptions&&);
```
`URLPatternInput&&`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,80 @@
+    using URLPatternInput = std::variant<String, URLPatternInit>;
+
+    static Ref<URLPattern> create(const URLPatternInput, const String&& baseURL, URLPatternOptions&&);
+    static Ref<URLPattern> create(std::optional<URLPatternInput>, URLPatternOptions&&);
```
`std::optional<URLPatternInput>&&`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,80 @@
+    static Ref<URLPattern> create(std::optional<URLPatternInput>, URLPatternOptions&&);
+    ~URLPattern();
+
+    ExceptionOr<bool> test(std::optional<URLPatternInput>, std::optional<String>&& baseURL) const;
```
`std::optional<URLPatternInput>&&`

cdumez replied:
Does it build if we use `String&&` for the second parameter? String has a null state in WebKit (String::isNull()) so we generally do not need to use `std::optional<String>`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -0,0 +1,80 @@
+
+    ExceptionOr<bool> test(std::optional<URLPatternInput>, std::optional<String>&& baseURL) const;
+
+    std::optional<URLPatternResult> exec(std::optional<URLPatternInput>&&, std::optional<String>&& baseURL) const;
```
Does it build if we use String&& for the second parameter? String has a null state in WebKit (String::isNull()) so we generally do not need to use std::optional<String>

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternInit.idl:
```diff
@@ -0,0 +1,41 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+ 
+ // https://urlpattern.spec.whatwg.org
```
Nit: we normally add the link to the specific interface implemented, for example:
```
// https://urlpattern.spec.whatwg.org/#dictdef-urlpatterninit
[
    JSGenerateToJSObject,
    JSGenerateToNativeObject
] dictionary URLPatternInit {
    USVString protocol;
    USVString username;
    USVString password;
    USVString hostname;
    USVString port;
    USVString pathname;
    USVString search;
    USVString hash;
    USVString baseURL;
};
```
Ditto for the other places.

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,68 @@
+
+WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(URLPattern);
+
+Ref<URLPattern> URLPattern::create(const URLPatternInput, const String&& baseURL, URLPatternOptions&&)
```
`const URLPatternInput&` or `URLPatternInput&&`

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -0,0 +1,68 @@
+    return Exception { ExceptionCode::NotSupportedError };
+}
+
+std::optional<URLPatternResult> URLPattern::exec(std::optional<URLPatternInput>&&, std::optional<String>&& baseURL) const
```
This should return `ExceptionOr<std::optional<URLPatternResult>>` (like the function above)

cdumez reviewed (approved):
LGTM

jelee53 committed:
Add stubs for URLPattern API
https://bugs.webkit.org/show_bug.cgi?id=280795
rdar://137166466

Reviewed by Chris Dumez.

URLPattern API spec https://urlpattern.spec.whatwg.org.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp: Added.
(WebCore::URLPattern::create):
(WebCore::URLPattern::test const):
(WebCore::URLPattern::exec const):
* Source/WebCore/Modules/url-pattern/URLPattern.h: Added.
* Source/WebCore/Modules/url-pattern/URLPattern.idl: Added.
* Source/WebCore/Modules/url-pattern/URLPatternInit.h: Added.
* Source/WebCore/Modules/url-pattern/URLPatternInit.idl: Added.
* Source/WebCore/Modules/url-pattern/URLPatternOptions.h: Added.
* Source/WebCore/Modules/url-pattern/URLPatternOptions.idl: Added.
* Source/WebCore/Modules/url-pattern/URLPatternResult.h: Added.
* Source/WebCore/Modules/url-pattern/URLPatternResult.idl: Added.
* Source/WebCore/Scripts/GenerateSettings.rb:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: https://commits.webkit.org/284734@main
